pol-core/bscript/object.cpp

```diff
@@ -1653,6 +1655,85 @@ BObjectImp* ObjArray::call_method_id( const int id, Executor& ex, bool /*forcebu
         return new BError( "array.randomentry() doesn't take parameters." );
     }
     break;
+  case MTH_FILTER:
+    if ( name_arr.empty() )
+    {
+      if ( ex.numParams() < 1 )
+        return new BError( "Invalid parameter type" );
+
+      BObjectImp* param0 = ex.getParamImp( 0, BObjectType::OTFuncRef );
+
+      if ( !param0 )
+        return new BError( "Invalid parameter type" );
+
+      // Some functions may allow optional arguments, so it is up to the caller
+      // of `makeContinuation` to validate argument count. If not handled,
+      // arguments will be shrunk and expanded (with uninit) as needed.
+      auto func = static_cast<BFunctionRef*>( param0 );
+      if ( func->numParams() != 1 )
+        return new BError( "Invalid parameter type" );
+
+      // If nothing to filter, return an empty array, since nothing to call the function with.
+      if ( ref_arr.empty() )
+        return new ObjArray();
+
+      // Arguments for user function call.
+      // Add the first element of the array to the call args.
+      BObjectRefVec args;
+      args.push_back( ref_arr.front() );
+
+      /*
+      The ContinuationCallback receives three arguments:
+
+      - `Executor&`
+      - `BContinuation* continuation`: The continuation, with methods to
+        handle the continuation (call the function again; finalize)
+      - `BObjectRef result`: The result of the user function call specified in
+        `makeContinuation`.
+
+      Return the `BObjectImp`:
+      - Call the user function again by returning a continuation
+      - Return something else to provide that value back to the script.
+      */
+      auto callback = [this, filteredObj = BObjectRef( new ObjArray ),
+                       processedObj = BObjectRef( new BLong( 1 ) )](
+                          Executor& ex, BContinuation* continuation,
+                          BObjectRef result ) -> BObjectImp*
+      {
+        auto filtered = static_cast<ObjArray*>( filteredObj->impptr() );
+        auto processed = static_cast<BLong*>( processedObj->impptr() );
```

Comment:
I think someday we should add to BObject
template <typename T>
T* ptr() {return static_cast<T*>(objimp.get());}
To not being forced to write everywhere such long lines.